Docker: Create Production Image

ardelato opened the pull request:
## Description

This creates a production instance of the Restarters app. We will have a separate Dockerfile and separate entrypoint file for the production image.

Utilizing this Dockerfile, we will push the image to ECR and then deploy the image with a Helm Chart and ArgoCD.

### CR Notes

We can definitely optimize the Dockerfile even more and potentially use a tool like https://github.com/slimtoolkit/slim to reduce the size of the image even more. Additionally, we can probably drop a lot of the PHP configurations and use the production config that already comes with [`php:fpm`](https://hub.docker.com/_/php). 

That said, for now this works and we can iterate on it.

---
 
qa_req 0 Utilizing this production image, I was able to deploy it with ArgoCD in our test cluster. That is also why non-docker changes were added to this pull, as the production image was not working as is.

ardelato committed:
feat(docker): add production Dockerfile and entrypoint script

Introduced a multi-stage Dockerfile for production builds, including PHP and NPM
dependencies, and optimized configurations for Laravel. Added an entrypoint script
to handle database connection and application initialization tasks. Updated
.dockerignore to exclude additional files and directories.

ardelato committed:
fix(composer): remove unused patches-file entry from composer.json

This was breaking the production docker image because we didn't copy the patch file json, even though we don't have anything in said patch file.

ardelato committed:
fix(commands): command structure and output handling

We should not have these output methods in the constructor method nor use
output-write. This ensures that the commands can be run in a non-interactive
manner like in the docker container.

ardelato committed:
fix(helpers): add namespace declaration to FixometerFile.php

Added the namespace declaration for the FixometerFile helper class to ensure
PSR-4 compatibility.

ardelato committed:
fix(helpers): set default values for emission ratios and displacement factor

Updated the LcaStats helper methods to include default values for emission ratios
and displacement factor when environment variables are not set. This ensures
consistent calculations and prevents potential errors due to missing environment
variables.

ardelato committed:
fix(wiki): implement conditional wiki integration handling

Added checks for the FEATURE__WIKI_INTEGRATION environment variable across
multiple components. This ensures that wiki-related functionalities are only
executed when the integration is enabled, preventing unnecessary operations
and potential errors when disabled. Updated UserController, event listeners,
and service providers accordingly.

mlahargou commented on app/Helpers/FixometerFile.php:
```diff
@@ -1,5 +1,7 @@
 <?php
 
+namespace App\Helpers;
```
Now that this is namespaced, do we need to update the usages of `FixometerFile`?

ardelato replied:
I didn't see any other errors or warnings. If errors do pop up for "production" then we can address it then.

ardelato replied:
Scratch that, its cause I didn't touch anything revolving image uploads.

We do need to update the usage.

ardelato replied:
Going to follow up on this issue in another pull.

mlahargou commented on app/Listeners/LogInToWiki.php:
```diff
@@ -37,6 +37,11 @@ public function __construct(Request $request, UserCreator $mediawikiUserCreator)
     public function handle(Login $event): void
     {
+        // Early return if wiki integration is disabled or dependencies are missing
+        if (env('FEATURE__WIKI_INTEGRATION') !== true || $this->wikiUserCreator === null) {
```
I think we can simplify a lot of these checks, but I'm fine being explicit. i.e.:

```php
        if (!env('FEATURE__WIKI_INTEGRATION') || !$this->wikiUserCreator) {

```

ardelato replied:
Addressed in [7e503d9](https://github.com/iFixit/restarters/pull/10/commits/7e503d94a8111e399e662aaa9abae1bfe58ee75d)

mlahargou commented on app/Providers/AppServiceProvider.php:
```diff
@@ -109,8 +109,14 @@ protected function registerEvents(): void
+        if (env('FEATURE__WIKI_INTEGRATION') === true) {
+            Event::listen(\App\Events\PasswordChanged::class, \App\Listeners\ChangeWikiPassword::class);
+        } else {
+            Event::listen(\App\Events\PasswordChanged::class, function() {
+                // Do nothing, wiki integration is disabled
+            });
```
I'm assuming something breaks if we just don't listen to the `PassowrdChanged` event when wiki integration is disabled?

ardelato replied:
I confirmed not registering an event listener for the PasswordChange event does not break anything.

Addressed in [9cb7d1f](https://github.com/iFixit/restarters/pull/10/commits/9cb7d1f4703b8e8cb26618acdd02fef0addcbc44)

mlahargou commented on app/Providers/MediawikiServiceProvider.php:
```diff
@@ -20,14 +27,30 @@ public function boot(): void
-        if (env('FEATURE__WIKI_INTEGRATION') == true) {
+        // Do not register any services if wiki integration is disabled
+        if (env('FEATURE__WIKI_INTEGRATION') !== true) {
+            // Register null implementations to avoid dependency injection errors
```
Eww, but sounds good.

ardelato replied:
Addressed in [a335c4f](https://github.com/iFixit/restarters/pull/10/commits/a335c4f1ef4b0b715318acd58276ab25404e394b)

Pretty much redid the registration logic.

mlahargou reviewed (approved):
CR 📱 with a few minor comments. I'll be honest, I'm pretty much trusting you on the dockerfile.

ardelato commented:
dev_block ⚡ on the FixometerFile namespace change.

ardelato committed:
fix(wiki): simplify conditional check for wiki integration

ardelato committed:
fix(wiki): remove no-op listener for password changed event when disabled

ardelato committed:
refactor(wiki): clean up MediawikiServiceProvider for conditional registration

Simplified the registration of Mediawiki-related services by introducing
dedicated methods for each service.

ardelato commented:
un_dev_block ⚡ looks like the image functionality (upload/edit) is more bugged than I thought. That said, its not a critical functionality so I will leave it as is for now.

mlahargou commented on app/Providers/MediawikiServiceProvider.php:
```diff
@@ -53,4 +81,47 @@ public function register(): void
+     *
+     * @return array
+     */
+    public function provides()
+    {
+        return [
+            MediawikiFactory::class,
+            UserCreator::class,
+            ActionApi::class,
+        ];
+    }
```
Just curious, why was this added?

ardelato replied:
Cause I didn't catch this 🙃. This isn't needed and was added as a debugging method.

I'll go ahead and remove it.

ardelato replied:
Addressed in [448e04f](https://github.com/iFixit/restarters/pull/10/commits/448e04f7b32e4b9453e1f892c5dbe2d0c226f3af)

mlahargou reviewed (approved):
CR 📱 I still think the namespace change will cause prod errors. But it sounds like you will resolve those separately.

ardelato committed:
refactor(discourse): remove unnecessary provides method

This isn't necessary and was used for debugging purposes.

mlahargou reviewed (approved):
CR 📱
